propulsion-hotel/Reactor/Handler.fs

```diff
@@ -0,0 +1,88 @@
+type Stats(log, statsInterval, stateInterval, ?logExternalStats) =
+    inherit Propulsion.Streams.Stats<Outcome>(log, statsInterval, stateInterval)
+
+    let mutable completed, deferred, failed, succeeded = 0, 0, 0, 0
+
+    override _.DumpStats() =
+        base.DumpStats()
+        if completed <> 0 || deferred <> 0 || failed <> 0 || succeeded <> 0 then
+            log.Information(" Completed {completed} Deferred {deferred} Failed {failed} Succeeded {succeeded}", completed, deferred, failed, succeeded)
+            completed <- 0; deferred <- 0; failed <- 0; succeeded <- 0
+        match logExternalStats with None -> () | Some f -> f Serilog.Log.Logger
+
+    override _.HandleOk res = res |> function
+        | Outcome.Completed -> completed <- completed + 1
+        | Outcome.Deferred -> deferred <- deferred + 1
+        | Outcome.Resolved successfully -> if successfully then succeeded <- succeeded + 1 else failed <- failed + 1
+    override _.Classify(exn) =
+        match exn with
+        | Equinox.DynamoStore.Exceptions.ProvisionedThroughputExceeded -> Propulsion.Streams.OutcomeKind.RateLimited
+        | x -> base.Classify x
+    override _.HandleExn(log, exn) =
+        log.Information(exn, "Unhandled")
+
+open Domain
+
+let private isReactionStream = function
+    | GroupCheckoutProcess.Category -> true
+    | _ -> false
+
+let private handleAction (guestStays : GuestStayAccount.Service) groupCheckoutId (act : GroupCheckoutProcess.Flow.Action) = async {
+    match act with
+    | GroupCheckoutProcess.Flow.Checkout pendingStays ->
+        let attempt stayId groupCheckoutId = async {
+            match! guestStays.GroupCheckout(stayId, groupCheckoutId) with
+            | GuestStayAccount.Decide.GroupCheckoutResult.Ok r -> return Choice1Of2 (stayId, r) 
+            | GuestStayAccount.Decide.GroupCheckoutResult.AlreadyCheckedOut -> return Choice2Of2 stayId } 
+        let! outcomes = seq { for x in pendingStays do attempt x groupCheckoutId } |> Async.parallelThrottled 5
+        let residuals, fails = outcomes |> Choice.partition id
+        return [ 
+            match residuals with
+            | [||] -> ()
+            | xs -> GroupCheckoutProcess.Events.GuestsMerged {| residuals = [| for stayId, amount in xs -> { stay = stayId; residual = amount } |] |}
+            match fails with
+            | [||] -> ()
+            | stayIds -> GroupCheckoutProcess.Events.GuestsFailed {| stays = stayIds |} ]
+    
+    | GroupCheckoutProcess.Flow.Ready _
+        // Nothing we can do other than wait for the Confirm to Come
+    | GroupCheckoutProcess.Flow.Finished ->
+        // No processing of any kind can happen after we reach this phase
+        return [] }
+
+let private handle handleAction (flow : GroupCheckoutProcess.Service) stream = async {
+    match stream with
+    | GroupCheckoutProcess.StreamName groupCheckoutId ->
+        let! ver' = flow.Pump(groupCheckoutId, handleAction groupCheckoutId)
```

Comment:
I found the `Pump` terminology hard to explain to a colleague. It sounds very infrastructural in nature but hangs on what should be a domain service.

Alternatives that don't have the same baggage: `Handle`, `Process`.

Reply:
I actually called it `process`... and that's a reserved word in F# for future use!
 I don't love pump. an there's a heck of a lot of `Handle` about.
 Will try to figure out a name for it, but the most important thing is that it makes sense in the context of the API of that `Service`
 🤔

Reply:
`flow.React` is conspicuously missing from the reactor 😊

Reply:
that probably would work well (wondering if the `flow` makes sense and/or whether I'm using it in enough / too many places too)

Reply:
maybe `handleAction` should be `handleReaction`, and `type Action` should be `type Reaction` (not sure that name makes sense in all contexts though - sometimes the `Action` is just an intrinsic piece of processing as part of a FSM processing chain rather than something that's separated, as implied by the _re_ in reaction) ?

Reply:
used React, handleReaction naming. Still considering `type Reaction` but have not pulled trigger....